src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs

```diff
@@ -2,57 +2,82 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+
 namespace System.Security.Cryptography
 {
     internal sealed partial class RandomNumberGeneratorImplementation : RandomNumberGenerator
     {
-        public override void GetNonZeroBytes(byte[] data)
+        public override void GetBytes(byte[] data)
         {
-            if (data == null)
-                throw new ArgumentNullException(nameof(data));
+            if (data == null) throw new ArgumentNullException(nameof(data));
+
+            GetBytesInternal(data, 0, data.Length);
+        }
 
-            GetBytes(data);
+        public override void GetBytes(byte[] data, int offset, int count)
+        {
+            if (data == null) throw new ArgumentNullException(nameof(data));
+            if (offset < 0) throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (offset + count > data.Length) throw new ArgumentException(SR.Argument_InvalidOffLen);
+
+            GetBytesInternal(data, offset, count);
+        }
 
-            int indexOfFirst0Byte = data.Length;
-            for (int i = 0; i < data.Length; i++)
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null) throw new ArgumentNullException(nameof(data));
+
+            int offset = 0;
+            while (offset < data.Length)
             {
-                if (data[i] == 0)
+                // Fill the remaining portion of the array with random bytes.
+                GetBytesInternal(data, offset, data.Length - offset);
+
+                // Find the first zero in the remaining portion.
+                int indexOfFirst0Byte = data.Length;
+                for (int i = offset; i < data.Length; i++)
                 {
-                    indexOfFirst0Byte = i;
-                    break;
+                    if (data[i] == 0)
+                    {
+                        indexOfFirst0Byte = i;
+                        break;
+                    }
                 }
-            }
 
-            for (int i = indexOfFirst0Byte; i < data.Length; i++)
-            {
-                if (data[i] != 0)
+                // If there were any zeros, shift down all non-zeros,
+                // leaving all of the zeros at the end.
```

Comment:
It doesn't swap the values, so { 3, 0, 8, 0, 192, 11 } turns into { 3, 8, 192, 11, 192, 11 }, then the last two values get overwritten with the next call to GetBytesInternal.

```C#
// If there were any zeros, shift down all non-zero values, and request new bytes.
//   offset = 0
//   filled data as { 3, 0, 8, 0, 192, 11 }
//   indexOfFirst0Byte = 1
//   { 3, 0, 8, 0, 192, 11 } (indexOfFirst0Byte = 1)
//   { 3, 8, 8, 0, 192, 11 } (indexOfFirst0Byte = 2)
//   { 3, 8, 192, 0, 192, 11 } (indexOfFirst0Byte = 3)
//   { 3, 8, 192, 11, 192, 11 } (indexOfFirst0Byte = 4)
//   offset = 4.
//   The last two bytes will be re-generated on the next pass.
```